doc: Snap commit-tracking guide (#18271)
canonical/lxd: doc/reference/snap_internals.md

---
relatedlinks: "[lxd-pkg-snap&#32;GitHub](https://github.com/canonical/lxd-pkg-snap)"
---

(snap-internals)=
# Snap internals

The LXD snap packaging is maintained in a separate git repository ([canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap)) from LXD itself ([canonical/lxd](https://github.com/canonical/lxd)). Launchpad builds the snap from a [third repository](https://code.launchpad.net/~lxd-snap/lxd/+git/lxd) on Launchpad which is pushed to by [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap)'s GitHub Actions.

> ## Tracking a commit in the LXD snap

Reviewer: (1) Avoid -ing verb forms, especially in headings. (2) Your main use case seems to be tracking bug fixes, so consider including "bug fix" in the heading. For example:

```suggestion
## Track a bug fix in the LXD snap
```

Given a bug report that has been fixed in LXD, this section shows where the fix has landed and where it has yet to land.

Consider [canonical/lxd#18023](https://github.com/canonical/lxd/issues/18023); this bug was introduced by commit [`d840004`](https://github.com/canonical/lxd/commit/d840004886b702b3bea15d4a1d6e4f32717a6b62) in LXD 3.0. The linked pull request points to commit [`c1e8ab4`](https://github.com/canonical/lxd/commit/c1e8ab4c33c217a200e59f27916fd8e1d49241e2) as the fix.

> Given this information, we can determine which snap channels have the fix and which ones don't.

Reviewer: A few comments here:

1. You have two sentences here that say similar things (lines 12 and 16). How about combining them? 
2. Let's make it really clear that as an **example**, we are going to determine whether or not a fix, commit `c1e8ab4`, to a bug (issue `18203`), has landed in a snap channel or not. In other words, let's use the word "example."
3. Let's minimize the number of commit hashes that we present to users. A user following the guide doesn't really need to know that `d840004` was the bug source. You could just say something along the lines of: the following examples show how to determine whether or not a fix to a [bug introduced in LXD 3.0](link to d840004) have landed in different snap channels.

Additionally, I would recommend declaring up front that the method for tracking the fix will vary depending on (1) risk level (latest vs candidate/stable) and (2) whether we are looking at `latest` or an LTS release.

Use `snap info` to show the currently available versions of LXD:

```{terminal}
:user: ubuntu
:host: ubuntu

snap info lxd

channels:
  5.21/stable:      5.21.4-aee7e08 2026-04-09 (38767)  123MB -
  5.21/candidate:   5.21.4-4189d16 2026-05-05 (39296)  123MB -
  5.21/beta:        ↑
  5.21/edge:        git-658877c    2026-05-07 (39329)  120MB -
  latest/stable:    6.7-d814d89    2026-04-03 (38768)  121MB -
  latest/candidate: 6.8-5a1a287    2026-05-06 (39313)  121MB -
  latest/beta:      ↑
  latest/edge:      git-3c2c6c6    2026-05-08 (39363)  119MB -
  6/stable:         6.7-d814d89    2026-04-03 (38768)  121MB -
  6/candidate:      6.8-5a1a287    2026-05-06 (39313)  121MB -
  6/beta:           ↑
  6/edge:           git-3c2c6c6    2026-05-08 (39363)  119MB -
```

> ### `latest/edge`

Reviewer: It seems to me that the "flowchart" for determining the method is first to ask "latest or LTS?" If "latest", then ask "edge" or feature release? If edge, then use `git merge-base`; if feature release, then use `git tag --contains`. If LTS, then use `git log ...`.

How about nesting headings as follows?

```
## Track a bug fix in the LXD snap
### `latest`
#### `edge` risk level
#### Feature release
### LTS releases
```

Author: I've rewritten the feature releases section to be more rigorous (using `git merge-base`); they're organized in order of complexity where each section builds on the last. That wasn't really clear in the first version, lmk what you think now.

Reviewer: That order of complexity is clear -- thanks!

> The snap version numbers for the `edge` {ref}`risk level <ref-snap-risk>` include the commit hash from [canonical/lxd](https://github.com/canonical/lxd) that was used to build that snap revision. To check if `latest/edge` and `6/edge` contain the fix:

Reviewer: I would provide the commit hashes in the body of the text here, to show the user what to look for. I also wonder if it might be simplest to only refer to `latest/edge` and then include a note at the end that `latest/edge` is the same as `6/edge`. For example:

```suggestion
The snap version numbers for the `edge` {ref}`risk level <ref-snap-risk>` include the commit hash from [canonical/lxd](https://github.com/canonical/lxd) that was used to build that snap revision. To check if `latest/edge` contains the bug fix, we need to determine if that fix (`c1e8ab41`) is an ancestor of that commit hash (`3c2c6c6`):
```

```{terminal}
:user: ubuntu
:host: ubuntu

git merge-base --is-ancestor c1e8ab4 3c2c6c6 && echo "c1e8ab4 reachable from 3c2c6c6"

c1e8ab4 reachable from 3c2c6c6
```

> This means that the fix is present in the `6/edge` and `latest/edge` channels.

Reviewer: Following up on my comment above about focusing on `latest/edge`:

````suggestion
This means that the fix is present in the `latest/edge` channels.
````

After this, you could include a "note" admonition pointing out that `latest` == `6` in this example.

(ref-internal-snap-feature-releases)=
### Feature releases

> The commit hash shown for the `candidate` and `stable` risk levels comes from [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap), so they can't be compared with the original fix commit.

Reviewer: As I noted above, I understood this example to be specific to `latest/candidate` and `latest/stable`. If so, I would use those prefixes, so that users don't confuse this with `5.21/stable` and `5.21/candidate`. You can also include another note that `latest` == `6`.

Author: I've done `s/the/all/` to make this more clear; this applies to all `candidate` and `stable` channels.

One strategy is to check [canonical/lxd](https://github.com/canonical/lxd) to determine which tagged version of LXD the fix is present in:

```{terminal}
:user: ubuntu
:host: ubuntu

git tag --contains c1e8ab4

lxd-6.8
```

This means that the fix is present in the `6/candidate` and `latest/candidate` channels.

> ```{note}

Reviewer: I would pull this information out of the note. If I understand you correctly, what you're saying is that if the `git tag --contains` approach fails, then you can try to check for a cherry-pick. I would also provide examples here. Show the user how to switch to the commit for the snap revision.

In other words, what you have is: one strategy (your phrasing) is to check for the tagged version; another strategy is to check for the original fix commit hash.

Author: Rewrote this section, lmk what you think

> Critical fixes may be cherry-picked at build time by `git` commands executed during the snap build process. [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) has branches for each snap channel. `git switch -d` to the commit for the snap revision you want to check and then search for the original fix commit hash in `snapcraft/snapcraft.yaml` to see if a cherry-pick has occurred for your change.

Author: Going to ignore this for now until we've had a human review...

```

### LTS releases

> Bugfixes are frequently backported to {ref}`LTS releases <ref-releases-snap>`; backports do not use the same commit hash and may not even have exactly the same code. Use `git log` to determine if [`c1e8ab4`](https://github.com/canonical/lxd/commit/c1e8ab4c33c217a200e59f27916fd8e1d49241e2) is present in the `5.21/edge` channel by checking the `stable-5.21` branch in [canonical/lxd](https://github.com/canonical/lxd) for commits containing the issue number or other pattern:

Reviewer: It seems like what you're looking for here is whether or not issue `18023` was addressed, rather than whether or not `c1e8ab4` is present. The example command below, for example, does not use the commit hash, but the issue number. Can you rewrite to clearly state that what we're trying to check is if the issue number appears in the git log?

```{terminal}
:user: ubuntu
:host: ubuntu

git log --grep="#18023" origin/stable-5.21

commit cce42965ad1e3b6a20995dc2c7a6a73ac2246944
Author: Thomas Parrott <thomas.parrott@canonical.com>
Date:   Wed Apr 15 16:04:40 2026 +0100

    lxd/networks: Only take networkCreateLock for external API requests

    This avoids deadlocks with the operation notification when multiple concurrent network creation requests arrive at different cluster members.

    Fixes #18023

    Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
```

[`c1e8ab4`](https://github.com/canonical/lxd/commit/c1e8ab4c33c217a200e59f27916fd8e1d49241e2) was backported as [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944).

> Use the same strategies as {ref}`feature releases <ref-internal-snap-feature-releases>` for identifying if [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944) is available in the `5.21/candidate` and `5.21/stable` channels.

Reviewer: ```suggestion
Use the same strategies as {ref}`feature releases <ref-internal-snap-feature-releases>` to determine if [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944) is available in the `5.21/candidate` and `5.21/stable` channels.
```

```{terminal}
:user: ubuntu
:host: ubuntu

snap info lxd

channels:
  5.21/stable:      5.21.4-aee7e08 2026-04-09 (38767)  123MB -
  5.21/candidate:   5.21.4-4189d16 2026-05-05 (39296)  123MB -
  5.21/beta:        ↑
  5.21/edge:        git-658877c    2026-05-07 (39329)  120MB -
  latest/stable:    6.7-d814d89    2026-04-03 (38768)  121MB -
  latest/candidate: 6.8-5a1a287    2026-05-06 (39313)  121MB -
  latest/beta:      ↑
  latest/edge:      git-3c2c6c6    2026-05-08 (39363)  119MB -
  6/stable:         6.7-d814d89    2026-04-03 (38768)  121MB -
  6/candidate:      6.8-5a1a287    2026-05-06 (39313)  121MB -
  6/beta:           ↑
  6/edge:           git-3c2c6c6    2026-05-08 (39363)  119MB -
```

The strategy used to track the fix will depend on both the snap's {ref}`risk level <ref-snap-risk>` (`edge` vs `candidate`/`stable`) and its {ref}`release type <ref-releases>`.

```{note}
The `latest/*` and `6/*` channels are equivalent in the above snap output; they point to the same snap revision. The strategies shown below for `latest/*` can be used unmodified for the current {ref}`feature release <ref-releases-feature>` channels (`6/*` in this case).
```

> ### `latest/edge`

Reviewer: ```suggestion
## Feature releases

### `latest/edge` risk level
```

The snap version numbers for all `edge` {ref}`risk levels <ref-snap-risk>` include the commit hash from [canonical/lxd](https://github.com/canonical/lxd) that was used to build that snap revision. To check if `latest/edge` contains the bug fix, clone [canonical/lxd](https://github.com/canonical/lxd) and check if the fix commit (`c1e8ab4`) is an ancestor of commit used to build the snap (`3c2c6c6`):

```{terminal}
:user: ubuntu
:host: ubuntu

git merge-base --is-ancestor c1e8ab4 3c2c6c6 && echo "c1e8ab4 reachable from 3c2c6c6"

c1e8ab4 reachable from 3c2c6c6
```

This means that the fix is present in the `latest/edge` channel.

(ref-internal-snap-feature-releases)=

> ### Feature releases (`latest/candidate` and `latest/stable`)

Reviewer: ```suggestion
### `latest/candidate` and `latest/stable` risk levels
```

The commit hash shown for all `candidate` and `stable` risk levels comes from [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap), so they can't be compared with the original fix commit.

> To identify the commit in [canonical/lxd](https://github.com/canonical/lxd) that corresponds to the [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) commit given for the snap revision, clone [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) and switch to the commit for `latest/candidate`: 

Reviewer: Maybe use the version number in the text to make this really clear:

```suggestion
To identify the commit in [canonical/lxd](https://github.com/canonical/lxd) that corresponds to the [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) commit given for the snap revision, clone [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) and switch to the commit in the `latest/candidate` version number (`6.8-5a1a287`): 
```

Author: This sentence is totally stammering; I've rewritten/incorporated it better.

```{terminal}
:user: ubuntu
:host: ubuntu

git switch --detach 5a1a287
```

Find the `source-commit` for that build:

```{terminal}
:user: ubuntu
:host: ubuntu

yq '.parts["lxd"]["source-commit"]' snapcraft.yaml

"84705553d17aeb8e15032611c321127a06c2f2ff"
```

> Then in [canonical/lxd](https://github.com/canonical/lxd), use `git merge-base` to identify if the fix is reachable:

Reviewer: ```suggestion
Then in [canonical/lxd](https://github.com/canonical/lxd), use `git merge-base` to check if the fix is reachable:
```

```{terminal}
:user: ubuntu
:host: ubuntu

git merge-base --is-ancestor c1e8ab4 8470555 && echo "c1e8ab4 reachable from 8470555"

c1e8ab4 reachable from 8470555
```

> This means that the fix is present in the `latest/candidate` channel. Identifying if the fix was present in `6.7-d814d89` is left as an exercise for the reader.

Reviewer: Rather than leaving it as an exercise, how about:

```suggestion
This means that the fix is present in the `latest/candidate` channel.
To check if the fix is present in the `latest/stable` channel, follow the procedure as above but use the commit hash from the `latest/stable` version number (`6.7-d814d89`).
```

```{important}
Critical fixes may be cherry-picked at build time by `git` commands executed during the snap build process. Use `yq '.parts["lxd"]["override-build"]' snapcraft.yaml` in [canonical/lxd-pkg-snap](https://github.com/canonical/lxd-pkg-snap) to identify if the fix was cherry-picked.
```

> ### LTS releases

Reviewer: ```suggestion
## LTS releases
```

Bugfixes are frequently backported to {ref}`LTS releases <ref-releases-snap>`; backports do not use the same commit hash and may not even have exactly the same code. Use `git log` to determine if the fix is present in the `5.21/edge` channel by checking the `stable-5.21` branch in [canonical/lxd](https://github.com/canonical/lxd) for commits containing the issue number (`#18023`) or the original commit hash (`c1e8ab4`):

```{terminal}
:user: ubuntu
:host: ubuntu

git log --grep="#18023" origin/stable-5.21

commit cce42965ad1e3b6a20995dc2c7a6a73ac2246944
Author: Thomas Parrott <thomas.parrott@canonical.com>
Date:   Wed Apr 15 16:04:40 2026 +0100

    lxd/networks: Only take networkCreateLock for external API requests

    This avoids deadlocks with the operation notification when multiple concurrent network creation requests arrive at different cluster members.

    Fixes #18023

    Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
```

[`c1e8ab4`](https://github.com/canonical/lxd/commit/c1e8ab4c33c217a200e59f27916fd8e1d49241e2) was backported as [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944).

> Use the same strategy as {ref}`feature releases <ref-internal-snap-feature-releases>` to determine if [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944) is available in the `5.21/candidate` and `5.21/stable` channels.

Reviewer: How about something like:

```suggestion
Follow the same strategy used for `stable` and `candidate` {ref}`feature releases <ref-internal-snap-feature-releases>` to determine if [`cce4296`](https://github.com/canonical/lxd/commit/cce42965ad1e3b6a20995dc2c7a6a73ac2246944) is available in the `5.21/candidate` and `5.21/stable` channels.
```
